autode/opt/coordinates/primitives.py

```diff
@@ -0,0 +1,222 @@
+    """Primitive internal coordinate"""
+
+    @abstractmethod
+    def __call__(self,
+                 x: 'autode.opt.coordinates.CartesianCoordinates'
+                 ) -> float:
+        """Return the value of this PIC given a set of cartesian coordinates"""
+
+    @abstractmethod
+    def derivative(self,
+                   i:         int,
+                   component: str,
+                   x:         'autode.opt.coordinates.CartesianCoordinates'
+                   ) -> float:
+        r"""
+        Calculate the derivative with respect to a cartesian coordinate
+
+        .. math::
+
+            \frac{dq}
+                  {d\boldsymbol{X}_{i, k}} {\Bigg\rvert}_{X=X0}
+
+        where :math:`q` is the primitive coordinate and :math:`\boldsymbol{X}`
+        are the cartesian coordinates.
+
+        -----------------------------------------------------------------------
+        Arguments:
+            i: Cartesian index to take the derivative with respect to;
+                0-N for N atoms
+
+            component: Cartesian component (x, y, z) to take the derivative
+                       with respect to. {'x', 'y', 'z'}
+
+            x: Cartesian coordinates
+
+        Returns:
+            (float): Derivative
+        """
+
+    @staticmethod
+    def _cart_component_to_idx(component: str) -> int:
+        """Convert the name of a Cartesian component (or direction) to an
+        index used for array indexing"""
+
+        if component == 'x':
+            return 0
+
+        if component == 'y':
+            return 1
+
+        if component == 'z':
+            return 2
+
+        else:
+            raise ValueError(f'Cannot convert component {component} to an '
+                             f'index. Must be one of x, y, z')
+
+    @abstractmethod
+    def __eq__(self, other):
+        """Comparison of two primitive coordinates"""
+
+
+class _DistanceFunction(Primitive, ABC):
+    """Function of a distance between two atoms"""
+
+    def __init__(self,
+                 idx_i: int,
+                 idx_j: int):
+        """
+        Function of a distance between a pair of atoms
+
+        .. math::
+
+            q = f(|\boldsymbol{X}_i - \boldsymbol{X}_j|)
+
+        for a set of cartesian coordinates :math:`\boldsymbol{X}`.
+        -----------------------------------------------------------------------
+        Arguments:
+            idx_i: Atom index
+
+            idx_j: Atom index
+        """
+
+        self.idx_i = int(idx_i)
+        self.idx_j = int(idx_j)
+
+    def __eq__(self, other) -> bool:
+        """Equality of two distance functions"""
+
+        return (isinstance(other, self.__class__)
+                and other._ordered_idxs == self._ordered_idxs)
+
+    @property
+    def _ordered_idxs(self) -> Tuple[int, int]:
+        """Indexes ordered by their value"""
+        i, j = self.idx_i, self.idx_j
+
+        return (i, j) if i < j else (j, i)
+
+
+class InverseDistance(_DistanceFunction):
```

Comment:
From what I can see in this module, I'm not sure having a `_DistanceFunction` class and subclasses is justified? Might be clearer to have distance calculation as top level functions

Reply:
It also makes sense too me, but I feel like there are more layers than necessary.

The only child class of `Primitive` I can find is `_DistanceFunction`. Unless you're absolutely certain you'll define more derived classes soon, I'd merge both `Primitive` and `_DistanceFunction` into an concrete class `Distance` that implements the distance coordinate, that is all methods `_cart_component_to_idx`, `ordered_idx` ,`eq`, `call` and `derivative`. Then `InvertDistance` inherits from `Distance` and reimplements both `derivative` and `call`. Maybe I'm missing something?